### PR TITLE
Mqtt implementation

### PR DESCRIPTION
**Pull request by nrjn**
m66_tcp.\* provides functions for basic operation like establish TCP connection, send data and receive data.
This will be further developed based on the applications (MQTT)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -34,6 +34,12 @@ provide(PACKAGE ubirch VERSION 1.0 BOARD ${BOARD} TARGETS board firmware)
 # add tests
 add_subdirectory(tests)
 
+# copy the config.h.template to config.h in case it is not there; it is ignored by .git!
```
Comment (thinkberg) [inline CMakeLists.txt]: Why did you add this here? It should be in tests/modem where the code that uses it is.

File: config.h.template [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#ifndef _CELL_H_
```
Comment (thinkberg) [inline config.h.template]: Same here, this file is a template belonging to the actual code.

File: firmware/CMakeLists.txt [hunk 3]
```diff
@@ -6,11 +6,11 @@ if (BOARD MATCHES "ubirch1r03")
     ubirch/m66/m66_http.c
     ubirch/m66/m66_ops.c
     ubirch/m66/m66_parser.c
+    ubirch/m66/m66_tcp.c
     )
   set(SDHC_SRCS
     ubirch/sdhc.c
-    ${FATFS_MIDDLEWARE_SRCS}
-    )
+    ${FATFS_MIDDLEWARE_SRCS})
```
Comment (thinkberg) [inline firmware/CMakeLists.txt]: Keep the single ')' on it's own line, makes adding and removing source simpler

File: tests/modem/modem.c [hunk 4]
```diff
@@ -0,0 +1,99 @@
+/**
+ * Identify the board.
```
Comment (thinkberg) [inline tests/modem/modem.c]: Fix comment header please.

File: tests/modem/modem.c [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  // INITIALIZATION
+  board_init();
+
+  // INITIALIZE CONSOLE (Tests debug uart pins!)
+  board_console_init(BOARD_DEBUG_BAUD);
+
+  // 100ms led blink, only works if setup for LED was correct
+  SysTick_Config(BOARD_SYSTICK_100MS / 10);
+
+  uint32_t uuid[4];
```
Comment (thinkberg) [inline tests/modem/modem.c]: remove parts of the code you don't really need for your case

File: firmware/ubirch/m66/m66_tcp.h [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// TODO Make the read buffer size configurable
+#ifndef MQTT_READ_BUFFER
+  #define MQTT_READ_BUFFER 255
+#endif
+
+bool modem_mqtt_connect(const char *apn, const char *user, const char *password, uint32_t timeout);
```
Comment (thinkberg) [inline firmware/ubirch/m66/m66_tcp.h]: This is confusing: m66_tcp.c, but inside is modem_mqtt_connect. Rename the functions to modem_tcp*

File: firmware/ubirch/m66/m66_tcp.h [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ```
+ */
+
+#ifndef UBIRCH_FIRMWARE_MQTT_H
```
Comment (thinkberg) [inline firmware/ubirch/m66/m66_tcp.h]: you are implementing TCP, not MQTT here, fix the define

File: CMakeLists.txt (hunk 1 above)
Comment (niranjanraoh) [inline CMakeLists.txt]: oops, I will change it

File: firmware/ubirch/m66/m66_tcp.c [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+#include "m66_parser.h"
+#include "ubirch/timer.h"
+#include <stdio.h>
+
+bool modem_mqtt_connect(const char *apn, const char *user, const char *password, uint32_t timeout) {
+  timer_set_timeout(timeout * 1000);
+
+  if (!modem_gprs_attach(apn, user, password, uTimer_Remaining)) return false;
+
+//  modem_send("AT+QIMUX=0");
```
Comment (thinkberg) [inline firmware/ubirch/m66/m66_tcp.c]: please add the comment what the commented out part does, or remove it if we will never use it

File: firmware/ubirch/m66/m66_tcp.c [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+//  if (!modem_expect_OK(uTimer_Remaining)) return false;
+
+//  Set QIMODE to 1, to establish TCP connection in transparent mode
+//  modem_send("AT+QIMODE=0");
+//  if (!modem_expect_OK(uTimer_Remaining)) return false;
+//  printf("MUX and MODE selected \r\n");
+
+  modem_send("AT+QIDNSIP=1");
+  if (!modem_expect_OK(uTimer_Remaining)) return false;
+  printf("DNS and not IP\r\n");
```
Comment (thinkberg) [inline firmware/ubirch/m66/m66_tcp.c]: use CSTDDEBUG() macro here for debug output

File: firmware/ubirch/m66/m66_tcp.c [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+  modem_send("AT+QISEND=%d", len);
+
+  if (modem_expect("> ", 300))
+  {
+    modem_send(data);
+  }
+  else
+  {
+    printf("we did not receive < \r\n");
```
Comment (thinkberg) [inline firmware/ubirch/m66/m66_tcp.c]: remove or use debug macro

File: firmware/ubirch/m66/m66_tcp.c [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+    return false;
+  }
+
+  if (!modem_expect("SEND OK", 300)) return false;
+
+  static char rx_buffer[] = {0};
+  static int rx_buffer_len = 0;
+
+  // Here it reads only one line
+  rx_buffer_len = modem_readline(rx_buffer, MQTT_READ_BUFFER, 2000);
```
Comment (thinkberg) [inline firmware/ubirch/m66/m66_tcp.c]: should we not read binary data from the modem (TCP connection)? reading by line will not work in most cases. Or is the Modem returning data line by line in an encoded format?

File: tests/modem/modem.c [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+
+  // 100ms led blink, only works if setup for LED was correct
+  SysTick_Config(BOARD_SYSTICK_100MS / 10);
+
+  // Initialize modem
+  modem_init();
+
+  if (!modem_enable())
+  {
+    printf("failed to enable modem\r\n");
```
Comment (thinkberg) [inline tests/modem/modem.c]: use PRINTF(), which are debug macros

File: tests/modem/modem.c [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+    return false;
+  }
+
+  if (!modem_tcp_connect(CELL_APN, CELL_USER, CELL_PWD, 20 * 5000))
+  {
+    printf("unable to connect \r\n");
+    return 0;
+  }
+
+  const char send_data[] = {"GET / HTTP/1.1\r\n\r\n"};
```
Comment (thinkberg) [inline tests/modem/modem.c]: this is not correct C code, you are initializing the array of characters with an array of array of characters.

{"something"} is a `char xxx[][]` or `char **xxx` 

initialization of a string should be:

```
const char send_data[] = "GET / HTTP/1.1\r\n\r\n";
```

File: firmware/ubirch/m66/m66_tcp.c (hunk 11 above)
Comment (niranjanraoh) [inline firmware/ubirch/m66/m66_tcp.c]: yes, read_binary makes sense if there are more than one line to read, which we have. I was planning on adding a function to read line by line - depending on the application

File: tests/modem/modem.c (hunk 13 above)
Comment (niranjanraoh) [inline tests/modem/modem.c]: oops, I will change it
